clang/test/SemaSYCL/intel-fpga-loops.cpp

```diff
@@ -468,6 +523,140 @@ void loop_count_control_dependent() {
+  constexpr int bar = 0;
+  [[intel::max_interleaving(bar + 2)]] for (int i = 0; i != 10; ++i) // OK
+      a[i] = 0;
+}
+
+void check_initiation_interval_expression() {
+  int a[10];
+  // Test that checks expression is not a constant expression.
+  // expected-note@+1{{declared here}}
+  int foo;
+  // expected-error@+2{{expression is not an integral constant expression}}
+  // expected-note@+1{{read of non-const variable 'foo' is not allowed in a constant expression}}
+  [[intel::initiation_interval(foo + 1)]] for (int i = 0; i != 10; ++i)
+       a[i] = 0;
+
+  // Test that checks expression is a constant expression.
+  constexpr int bar = 0;
+  [[intel::initiation_interval(bar + 2)]] for (int i = 0; i != 10; ++i) // OK
+      a[i] = 0;
+}
+
+void check_speculated_iterations_expression() {
+  int a[10];
+  // Test that checks expression is not a constant expression.
+  // expected-note@+1{{declared here}}
+  int foo;
+  // expected-error@+2{{expression is not an integral constant expression}}
+  // expected-note@+1{{read of non-const variable 'foo' is not allowed in a constant expression}}
+  [[intel::speculated_iterations(foo + 1)]] for (int i = 0; i != 10; ++i)
+       a[i] = 0;
+
+  // Test that checks expression is a constant expression.
+  constexpr int bar = 0;
+  [[intel::speculated_iterations(bar + 2)]] for (int i = 0; i != 10; ++i) // OK
+      a[i] = 0;
+}
+
+void check_loop_coalesce_expression() {
+  int a[10];
+  // Test that checks expression is not a constant expression.
+  // expected-note@+1{{declared here}}
+  int foo;
+  // expected-error@+2{{expression is not an integral constant expression}}
+  // expected-note@+1{{read of non-const variable 'foo' is not allowed in a constant expression}}
+  [[intel::loop_coalesce(foo + 1)]] for (int i = 0; i != 10; ++i)
+       a[i] = 0;
+
+  // Test that checks expression is a constant expression.
+  constexpr int bar = 0;
+  [[intel::loop_coalesce(bar + 2)]] for (int i = 0; i != 10; ++i) // OK
+      a[i] = 0;
+}
+
+void check_loop_count_expression() {
+  int a[10];
+
+  // Test that checks expression is not a constant expression.
+  int foo; // expected-note {{declared here}}
+  // expected-error@+2{{expression is not an integral constant expression}}
+  // expected-note@+1{{read of non-const variable 'foo' is not allowed in a constant expression}}
+  [[intel::loop_count_max(foo + 1)]] for (int i = 0; i != 10; ++i)
+       a[i] = 0;
+
+  // Test that checks expression is a constant expression.
+  constexpr int bar = 0;
+  [[intel::loop_count_max(bar + 2)]] for (int i = 0; i != 10; ++i) // OK
+      a[i] = 0;
+}
+
+// Test that checks wrong template instantiation and ensures that the type
+// is checked properly when instantiating from the template definition.
+struct S {};
+template <typename Ty>
+void check_loop_attr_template_instantiation() {
+  int a[10];
+
+  // expected-error@+2 {{integral constant expression must have integral or unscoped enumeration type, not 'S'}}
```

Comment:
Sorry, can you explain this test. Why does instantiating with `Ty{}` throw error about `S`

Reply:
// This test checks the type here when instantiating from the template definition.


```
struct S {};
template <typename Ty>
void check_loop_attr_template_instantiation() {
  int a[10];
  // expected-error@+2 {{integral constant expression must have integral or unscoped enumeration type, not 'S'}}
  // expected-error@+1 {{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
  [[intel::loop_coalesce(Ty{})]] for (int i = 0; i != 10; ++i)
      a[i] = 0;
}

int main() {
  check_loop_attr_template_instantiation<S>(); //expected-note{{in instantiation of function template specialization 'check_loop_attr_template_instantiation<S>' requested here}}
}
```

Reply:
Thanks for clarification!